jitstats.lua

```diff
@@ -0,0 +1,462 @@
+  
+  -- Recording calls/returns.
+  {name = "BADTYPE", msg = "bad argument type"},
+  {name = "CJITOFF", msg = "JIT compilation disabled for function"},
+  {name = "CUNROLL", msg = "call unroll limit reached"},
+  {name = "DOWNREC", msg = "down-recursion, restarting"},
+  {name = "NYIFFU",  msg = "NYI: unsupported variant of FastFunc %s"},
+  {name = "NYIRETL", msg = "NYI: return to lower frame"},
+  
+  -- Recording indexed load/store.
+  {name = "STORENN", msg = "store with nil or NaN key"},
+  {name = "NOMM",    msg = "missing metamethod"},
+  {name = "IDXLOOP", msg = "looping index lookup"},
+  {name = "NYITMIX", msg = "NYI: mixed sparse/dense table"},
+  
+  -- Recording C data operations. */
+  {name = "NOCACHE", msg = "symbol not in cache"},
+  {name = "NYICONV", msg = "NYI: unsupported C type conversion"},
+  {name = "NYICALL", msg = "NYI: unsupported C function type"},
+  
+  -- Optimizations.
+  {name = "GFAIL",   msg = "guard would always fail"},
+  {name = "PHIOV",   msg = "too many PHIs"},
+  {name = "TYPEINS", msg = "persistent type instability"},
+  
+  -- Assembler.
+  {name = "MCODEAL", msg = "failed to allocate mcode memory"},
+  {name = "MCODEOV", msg = "machine code too long"},
+  {name = "MCODELM", msg = "hit mcode limit (retrying)"},
+  {name = "SPILLOV", msg = "too many spill slots"},
+  {name = "BADRA",   msg = "inconsistent register allocation"},
+  {name = "NYIIR",   msg = "NYI: cannot assemble IR instruction %d"},
+  {name = "NYIPHI",  msg = "NYI: PHI shuffling too complex"},
+  {name = "NYICOAL", msg = "NYI: register coalescing too complex"},
+}
+
+local traceerr, traceerr_fromid = {}, table.new(#vmdef.traceerr, 0)
+
+for i, evt in pairs(traceerror_info) do
+    local id = traceerr_lookup[evt.msg]
+
+    if id then
+        traceerr[evt.name] = id
+        traceerr_fromid[id] = evt.name
+    end
+end
+
+local TRERR_BLACKL, TRERR_LLEAVE, TRERR_LUNROLL, TRERR_NYIBC = traceerr.BLACKL, traceerr.LLEAVE, traceerr.LUNROLL, traceerr.NYIBC
+
+local function printf(fmt, ...)
+    if select("#", ...) == 0 then
+        print(fmt)
+    else
+        print(string.format(fmt, ...))
+    end
+end
+
+local debugprint = function () end
+
+-- Stolen from dump.lua
+local function fmtfunc(func, pc)
+    local fi = funcinfo(func, pc)
+    if fi.loc then
+        return fi.loc
+    elseif fi.ffid then
+        return vmdef.ffnames[fi.ffid]
+    elseif fi.addr then
+        return ("C:%x"):format(fi.addr)
+    else
+        return "(?)"
+    end
+end
+
+-- Format trace error message.  Stolen from dump.lua
+local function fmterr(err, info)
+    if type(err) == "number" then
+        if type(info) == "function" then 
+            info = fmtfunc(info) 
+        end
+        err = vmdef.traceerr[err]:format(info)
+    end
+    return err
+end
+
+local jitstats = {
+    isrunning = false,
+    debug = false,
+    stats = {}
+}
+
+local stats = jitstats.stats
+
+local vmevent = {}
+
+local curr_parentid, curr_stitched
+local start_pc, start_func
```

Comment:
these variables are used only on `vmevent.start/stop/abort()`, wrap in a `do...end` to limit scope?

Reply:
I've never really needed to scope variables like that